### DISCUSSION TITLE
Mustaching Mustached html variables from python

### DISCUSSION THREAD

**Discussion by JamesGiaquinto** [PrairieLearn/PrairieLearn#6030] [Q&A] [answered]
Hi, I'm using mustaching, but I'm trying to call html saved in python that has mustaches of it's own.
```
<hmt>
   {{{params.random_display}}}

.py
   a,b,c = random.sample(range(1,50),3)
   data['params']['a'] = a
   data['params']['b'] = b
   data['params']['c'] = c
   data['params']['random_display'] = random.choice([
      '<p>forward {{params.a}}, {{params.b}}, {{params.c}}.</p>',
      '<p>backward {{params.c}}, {{params.b}}, {{params.a}}
   ])

output:
   forward {{params.a}}, {{params.b}}, {{params.c}}.
```
I'm using the triple mustache in html. I'm I doing something wrong? Thanks

Thank you, Jonatan for a great solution.
[Jonatan][43 minutes ago](https://prairielearn.slack.com)
You can use Python f-strings:

    data['params']['random_display'] = random.choice([
       f'<p>forward {a}, {b}, {c}.</p>',
       f'<p>backward {c}, {b}, {a}'
    ])

**Comment by JamesGiaquinto** [comment]
Ran into this error when trying to reference hashable variable:
SyntaxError: f-string expression part cannot include '#'

```
data['params']['some_set'] = [
  {
       'add':  5+i,
       'multi': 5*i
  }
   for i in random.sample(range(1,50), k=5)
]
data['params']['random_display'] = random.choice([
    '<p>Not the hashable set.</p>',
    '{{#params.some_set}}<p>Adding 5 is {{add}}, while multiplying is {{multi}}.</p>{{/params.some_set}}'

```
Hopefully, the outcome in html would be 5 lines of a random number (1,50) added and multiplied.
With f'string' I was able to substitute single variables. Hashing got complicated. Any tips?

**Reply by jonatanschroeder** [reply]
Try:

```
''.join([f'<p>Adding 5 is {e["add"]}, while multiplying is {e["multi"]}.</p>' for e in data["params"]["some_set"]])
```

**Reply by echuber2** [reply]
Just to elaborate: the `{{#params.some_set}}` ... `{{/params.some_set}}` Mustache part isn't expanded in the Python code before the random choice would be made. The Mustache processing steps run later on the assembled question page during rendering. In contrast, Python f-strings are a native Python thing and the templates in f-strings do get expanded when the string is evaluated in Python. So, using a native Python loop as Jonatan showed is a good solution, instead of trying to use a Mustache loop.

**Comment by jonatanschroeder** [comment, marked as answer]
Try:

```
''.join([f'<p>Adding 5 is {e["add"]}, while multiplying is {e["multi"]}.</p>' for e in data["params"]["some_set"]])
```

**Reply by JamesGiaquinto** [reply]
Heck, Yes! That's what I'm talking about. Everyone here rocks!
It's the Stem and Leaf Plot that is repeating
![image](https://user-images.githubusercontent.com/85510847/180131019-741adbf8-feee-4b87-b41c-f19507a9f387.png)
